lib/ecto/query/builder/from.ex

```diff
@@ -103,13 +103,22 @@ defmodule Ecto.Query.Builder.From do
   @doc """
   The callback applied by `build/2` to build the query.
   """
-  @spec apply(Ecto.Queryable.t, non_neg_integer) :: Ecto.Query.t
-  def apply(query, binds) do
-    query = Ecto.Queryable.to_query(query)
+  @spec apply(Ecto.Queryable.t, non_neg_integer, atom) :: Ecto.Query.t
+  def apply(query, binds, as) do
+    query =
+      query
+      |> Ecto.Queryable.to_query()
+      |> maybe_apply_as(as)
+
     check_binds(query, binds)
     query
   end
 
+  defp maybe_apply_as(query, nil), do: query
+  defp maybe_apply_as(%{from: from, aliases: aliases} = query, as) do
+    %{query | aliases: Map.put(aliases, as, 0), from: %{from | as: as}}
```

Comment:
We have two issues here:

  1. What happens if the alias already exists?
  2. What happens if the `from` already has an `as`?

We need to check for those two cases and raise accordingly. :)

Reply:
@josevalim Fixed in https://github.com/elixir-ecto/ecto/pull/2522/commits/77f499588a618f40a5fb212880765727a974c5db. Don't know what I was thinking.